### PR TITLE
build: split depends Qt into native and target builds

### LINKED ISSUE #18536: build: Fail to build Qt in depends on ARM 32bit
Having the low-end system [ODROID-HC1](https://www.hardkernel.com/shop/odroid-hc1-home-cloud-one/):
```
$ uname -srm
Linux 4.14.173-173 armv7l
$ lsb_release -ds
Ubuntu 18.04.4 LTS
$ /usr/bin/make --version | head -2
GNU Make 4.1
Built for arm-unknown-linux-gnueabihf
```

Trying to build depends:
```
$ make -C depends
...
Configure summary:

Build type: linux-g++ (arm, CPU features: <none>)
Configuration: enable_new_dtags largefile precompile_header silent release c++11 dbus no-qml-debug reduce_exports release_tools static stl
Build options:
  Mode ................................... release; optimized tools
  Optimize release build for size ........ no
  Building shared libraries .............. no
  Using C++ standard ..................... C++11
  Using ccache ........................... no
  Using gold linker ...................... no
  Using new DTAGS ........................ yes
  Using precompiled headers .............. yes
  Using LTCG ............................. no
  Target compiler supports:
    NEON ................................. no
  Build parts ............................ libs tools
Qt modules and options:
  Qt Concurrent .......................... no
  Qt D-Bus ............................... yes
  Qt D-Bus directly linked to libdbus .... no
  Qt Gui ................................. yes
  Qt Network ............................. yes
  Qt Sql ................................. no
  Qt Testlib ............................. yes
  Qt Widgets ............................. yes
  Qt Xml ................................. no
Support enabled for:
  Using pkg-config ....................... yes
  QML debugging .......................... no
  udev ................................... no
  Using system zlib ...................... yes
Qt Core:
  DoubleConversion ....................... yes
    Using system DoubleConversion ........ no
  GLib ................................... no
  iconv .................................. no
  ICU .................................... no
  Logging backends:
    journald ............................. no
    syslog ............................... no
    slog2 ................................ no
  Using system PCRE2 ..................... no
Qt Network:
  getaddrinfo() .......................... yes
  getifaddrs() ........................... yes
  IPv6 ifname ............................ yes
  libproxy ............................... no
  OpenSSL ................................ yes
    Qt directly linked to OpenSSL ........ yes
  SCTP ................................... no
  Use system proxies ..................... yes
Qt Gui:
  Accessibility .......................... yes
  FreeType ............................... yes
    Using system FreeType ................ yes
  HarfBuzz ............................... yes
    Using system HarfBuzz ................ no
  Fontconfig ............................. yes
  Image formats:
    GIF .................................. no
    ICO .................................. no
    JPEG ................................. no
      Using system libjpeg ............... no
    PNG .................................. yes
      Using system libpng ................ no
  EGL .................................... no
  OpenVG ................................. no
  OpenGL:
    Desktop OpenGL ....................... no
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
  Session Management ..................... no
Features used by QPA backends:
  evdev .................................. yes
  libinput ............................... no
  INTEGRITY HID .......................... no
  mtdev .................................. no
  tslib .................................. no
  xkbcommon-evdev ........................ no
QPA backends:
  DirectFB ............................... no
  EGLFS .................................. no
  LinuxFB ................................ no
  VNC .................................... no
  Mir client ............................. no
  X11:
    Using system-provided XCB libraries .. no
    EGL on X11 ........................... no
    Xinput2 .............................. no
    XCB XKB .............................. yes
    XLib ................................. no
    XCB render ........................... yes
    XCB GLX .............................. no
    XCB Xlib ............................. no
    Using system-provided xkbcommon ...... no
Qt Widgets:
  GTK+ ................................... no
  Styles ................................. Fusion Windows
Qt PrintSupport:
  CUPS ................................... no
Qt Sql:
  DB2 (IBM) .............................. no
  InterBase .............................. no
  MySql .................................. no
  OCI (Oracle) ........................... no
  ODBC ................................... no
  PostgreSQL ............................. no
  SQLite2 ................................ no
  SQLite ................................. no
    Using system provided SQLite ......... no
  TDS (Sybase) ........................... no
...
Project ERROR: Unknown module(s) in QT: xml
funcs.mk:251: recipe for target '/home/hebasto/bitcoin/depends/work/build/armv7l-unknown-linux-gnueabihf/qt/5.9.8-32c67940323/qtbase/.stamp_configured' failed
```

With `NO_QT=1` depends are built successfully.

### PR DESCRIPTION
**Pull request by fanquake**
This splits our Qt build in depends into two parts. The first builds `qmake`, qt tools (`uic`, `moc`, `rcc` etc) as well as translation tools (`lconvert`, `lupdate` etc). The second builds the libs we want for `bitcoin-qt`.

Splitting the build in this way has a few advantages. For example, we can enable the xml module while building the tools, and ideally fix issues like #14648 or #18536 without needing changes like #21589 or #21420, or having to build it alongside our libs.

This has been tested a little bit. GUIX builds are working for macOS and Windows.

Some testing still needs to be done in regards to passing compiler flags, DEBUG=1 builds etc. No doubt some improvements can still be made to what is currently here.

Fixes #18536.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK. The same idea (a separated making of tools) is used for other dependencies.

Need more investigation to find out if this approach is compatible with adding additional Qt modules.

**Comment by hebasto** [comment]
I think the reason of the failure on Centos 8 is wrong architecture the `qmake` is configured for -- `x86_64` instead of `i386`.

**Comment by hebasto** [comment]
I'm able to `make -C depends qt_configured HOST=i686-pc-linux-gnu` with the correct architecture detection:
```
...
Configure summary:

Building on: linux-g++ (x86_64, CPU features: mmx sse sse2)
Building for: linux-g++-32 (i386, CPU features: <none>)
...
```
with the following patch:
```diff
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -111,10 +111,6 @@ $(package)_config_opts += -no-feature-vulkan
 $(package)_config_opts += -no-feature-wizard
 $(package)_config_opts += -no-feature-xml
 
-$(package)_config_opts += "QMAKE_CFLAGS = '$($(package)_cflags) $($(package)_cppflags)'"
-$(package)_config_opts += "QMAKE_CXXFLAGS = '$($(package)_cflags) $($(package)_cppflags)'"
-$(package)_config_opts += "QMAKE_LFLAGS = '$($(package)_ldflags)'"
-
 $(package)_config_opts_darwin = -no-feature-corewlan
 $(package)_config_opts_darwin += QMAKE_MACOSX_DEPLOYMENT_TARGET=$(OSX_MIN_VERSION)
 
@@ -217,6 +213,9 @@ define $(package)_preprocess_cmds
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
   cp -r qtbase/mkspecs/linux-arm-gnueabi-g++ qtbase/mkspecs/bitcoin-linux-g++ && \
   sed -i.old "s/arm-linux-gnueabi-/$(host)-/g" qtbase/mkspecs/bitcoin-linux-g++/qmake.conf && \
+  echo "!host_build: QMAKE_CFLAGS     += $($(package)_cflags) $($(package)_cppflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
+  echo "!host_build: QMAKE_CXXFLAGS   += $($(package)_cxxflags) $($(package)_cppflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
+  echo "!host_build: QMAKE_LFLAGS     += $($(package)_ldflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
   sed -i.old "s|QMAKE_CC                = \$$$$\$$$${CROSS_COMPILE}clang|QMAKE_CC                = $($(package)_cc)|" qtbase/mkspecs/common/clang.conf && \
   sed -i.old "s|QMAKE_CXX               = \$$$$\$$$${CROSS_COMPILE}clang++|QMAKE_CXX               = $($(package)_cxx)|" qtbase/mkspecs/common/clang.conf && \
   sed -i.old "s/LIBRARY_PATH/(CROSS_)?\0/g" qtbase/mkspecs/features/toolchain.prf
```

**Comment by hebasto** [comment]
The branch with applied [fixup](https://github.com/bitcoin/bitcoin/pull/22142#issuecomment-856926951) is [here](https://github.com/hebasto/bitcoin/commits/210608-qtbase-5).
CI -- https://cirrus-ci.com/build/4834341327994880

**Comment by hebasto** [comment]
> The branch with applied [fixup](https://github.com/bitcoin/bitcoin/pull/22142#issuecomment-856926951) is [here](https://github.com/hebasto/bitcoin/commits/210608-qtbase-5).

Tested on Armbian 21.02.3 -- #18536 is fixed, and #21589 could be dropped.

File: depends/packages/native_qtbase.mk [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+define $(package)_set_vars
+$(package)_config_opts_release = -release
+$(package)_config_opts_release += -silent
+$(package)_config_opts_debug = -debug
+$(package)_config_opts_debug += -optimized-tools
+$(package)_config_opts += -bindir $(build_prefix)/bin
+$(package)_config_opts += -c++std c++1z
+$(package)_config_opts += -confirm-license
+$(package)_config_opts += -hostprefix $(build_prefix)
+$(package)_config_opts += -hostdatadir $(build_prefix)/share
```
Comment (hebasto) [inline depends/packages/native_qtbase.mk:23]: This line seems unneeded. Verified with 9046de8a4cbc3899fed9eae084115f423e7ac5bd (from #21995).

File: depends/packages/native_qtbase.mk [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+define $(package)_fetch_cmds
+$(call fetch_file,$(package),$($(package)_download_path),$($(package)_download_file),$($(package)_file_name),$($(package)_sha256_hash)) && \
+$(call fetch_file,$(package),$($(package)_download_path),$($(package)_qttools_file_name),$($(package)_qttools_file_name),$($(package)_qttools_sha256_hash))
+endef
+
+define $(package)_extract_cmds
+  mkdir -p $($(package)_extract_dir) && \
+  echo "$($(package)_sha256_hash)  $($(package)_source)" > $($(package)_extract_dir)/.$($(package)_file_name).hash && \
+  echo "$($(package)_qttools_sha256_hash)  $($(package)_source_dir)/$($(package)_qttools_file_name)" >> $($(package)_extract_dir)/.$($(package)_file_name).hash && \
```
Comment (hebasto) [inline depends/packages/native_qtbase.mk:143]: style nit, extra spaces:
```suggestion
  echo "$($(package)_sha256_hash) $($(package)_source)" > $($(package)_extract_dir)/.$($(package)_file_name).hash && \
  echo "$($(package)_qttools_sha256_hash) $($(package)_source_dir)/$($(package)_qttools_file_name)" >> $($(package)_extract_dir)/.$($(package)_file_name).hash && \
```

**Comment by hebasto** [comment]
The `linguist` stuff could be simplified:
```diff
--- a/depends/packages/native_qtbase.mk
+++ b/depends/packages/native_qtbase.mk
@@ -162,9 +162,7 @@ define $(package)_config_cmds
   qtbase/bin/qmake -o qtbase/src/tools/qfloat16-tables/Makefile qtbase/src/tools/qfloat16-tables/qfloat16-tables.pro && \
   qtbase/bin/qmake -o qtbase/src/tools/qvkgen/Makefile qtbase/src/tools/qvkgen/qvkgen.pro && \
   qtbase/bin/qmake -o qtbase/qmake/Makefile qtbase/qmake/qmake-aux.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lrelease/Makefile qttools/src/linguist/lrelease/lrelease.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lupdate/Makefile qttools/src/linguist/lupdate/lupdate.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lconvert/Makefile qttools/src/linguist/lconvert/lconvert.pro
+  qtbase/bin/qmake -o qttools/src/linguist/Makefile qttools/src/linguist/linguist.pro
 endef
 
 define $(package)_build_cmds
@@ -174,9 +172,7 @@ define $(package)_build_cmds
   $(MAKE) -C qtbase/src/tools/qvkgen && \
   $(MAKE) -C qtbase/src/tools/rcc && \
   $(MAKE) -C qtbase/src/tools/uic && \
-  $(MAKE) -C qttools/src/linguist/lrelease && \
-  $(MAKE) -C qttools/src/linguist/lupdate && \
-  $(MAKE) -C qttools/src/linguist/lconvert
+  $(MAKE) -C qttools/src/linguist
 endef
 
 define $(package)_stage_cmds
```

Verified with 9046de8a4cbc3899fed9eae084115f423e7ac5bd (from #21995).

File: depends/packages/native_qtbase.mk (hunk 2 above)
Comment (fanquake) [reply depends/packages/native_qtbase.mk:143]: Addressed in next push.

File: depends/packages/native_qtbase.mk (hunk 1 above)
Comment (fanquake) [reply depends/packages/native_qtbase.mk:23]: Addressed in next push.

**Comment by fanquake** [comment]
> I'm able to make -C depends qt_configured HOST=i686-pc-linux-gnu with the correct architecture detection:
> The linguist stuff could be simplified:

Thanks, I've addressed both of these points. Also addressed the comments you left inline and rebased for #22174. I think we can merge #22186 before this. I'm not really happy with how qt patches end up duplicated, but we should be able to fix that.

File: depends/packages/qt.mk [hunk 3]
```diff
@@ -179,48 +167,36 @@ endef
...
 #
-# 3. Create a macOS-Clang-Linux mkspec using our mac-qmake.conf.
-#
-# 4. After making a copy of the mkspec for the linux-arm-gnueabi host, named
+# 3. After making a copy of the mkspec for the linux-arm-gnueabi host, named
 # bitcoin-linux-g++, replace instances of linux-arm-gnueabi with $(host). This
 # way we can generically support hosts like riscv64-linux-gnu, which Qt doesn't
 # ship a mkspec for. See it's usage in config_opts_* above.
 #
-# 5. Put our C, CXX and LD FLAGS into gcc-base.conf. Only used for non-host builds.
```
Comment (fanquake) [inline depends/packages/qt.mk:214]: #5 needs to be restored here.

**Comment by hebasto** [comment]
> Splitting the build in this way has a few advantages. For example, we can enable the xml module while building the tools, and ideally fix issues like #14648 or #18536 without needing changes like #21589 or #21420, or having to build it alongside our libs.

It is true that avoiding intrusive fixes to the build system--i.e., patching Qt code base--makes it more maintainable and more comprehensible. That is why Concept ACKed this PR.

But considering different approaches, I still prefer the #20641. It is better than this one for the following reasons:
- no `qmake` building duplication
- no patch duplication
- using Qt "blessed" approach, that mimics building from the single source tree that includes all modules
- compatibility with possible including other modules in the future (e.g., QtQml for #16883, or QtMultimedia for Qr-code scanning)

**Comment by fanquake** [comment]
> compatibility with possible including other modules in the future

Is this approach actually incompatible with doing that?

**Comment by hebasto** [comment]
> > compatibility with possible including other modules in the future
> 
> Is this approach actually incompatible with doing that?

If new modules depends on core and tools are built in the `native_qtbase.mk` only -- that is ok.
But if they have mutual dependencies, this approach could fail (without additional hacks).

**Comment by fanquake** [comment]
Closing this in favour of #20641 for now.

**Pull request closed without merging**